### PR TITLE
Ensure mtime of site packages is updated during wheel installation

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/uv/issues/2530

### CODE REVIEW

File: crates/install-wheel-rs/src/linker.rs [hunk 1]
```diff
@@ -249,6 +252,29 @@ fn clone_wheel_files(
...
+        let mtime_file = site_packages
+            .as_ref()
+            .join(format!("uv-mtime-{}", hasher.finish()));
+
+        // Create and remove the file
+        fs_err::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(mtime_file.clone())?;
+        fs_err::remove_file(mtime_file)?;
```
Comment (zanieb) [inline crates/install-wheel-rs/src/linker.rs:275]: Arguably we could avoid failing on these, but it seems nice to know if we're leaving them around due to some failure.

File: crates/install-wheel-rs/src/linker.rs [hunk 2]
```diff
@@ -249,6 +252,29 @@ fn clone_wheel_files(
     let mut count = 0usize;
     let mut attempt = Attempt::default();
 
+    if cfg!(target_os = "macos") {
+        // On macOS, the directory mtime is not updated when cloning occurs and the mtime is used
+        // by CPython's import mechanisms to determine if it should look for new packages in a directory.
+        // Here, we force the mtime to be updated to ensure that packages are importable without
+        // manual cache invalidation.
```
Comment (zanieb) [inline crates/install-wheel-rs/src/linker.rs:259]: We should probably add a test that asserts the mtime changes on install

File: crates/install-wheel-rs/src/linker.rs [hunk 3]
```diff
@@ -249,6 +252,29 @@ fn clone_wheel_files(
...
+        //
+        // <https://github.com/python/cpython/blob/8336cb2b6f428246803b02a4e97fce49d0bb1e09/Lib/importlib/_bootstrap_external.py#L1601>
+
+        // Construct a unique file name
+        let mut hasher = FxHasher::default();
+        hasher.write(wheel.as_ref().as_os_str().as_bytes());
+        let mtime_file = site_packages
+            .as_ref()
+            .join(format!("uv-mtime-{}", hasher.finish()));
```
Comment (zanieb) [inline crates/install-wheel-rs/src/linker.rs:268]: We could also use a `SmallRng` here, I don't have strong feelings.

File: crates/install-wheel-rs/src/linker.rs (hunk 1 above)
Comment (zanieb) [reply crates/install-wheel-rs/src/linker.rs]: I'm looking into that next yeah, https://doc.rust-lang.org/std/fs/struct.File.html#method.set_modified also

Comment (charliermarsh) [reply crates/install-wheel-rs/src/linker.rs]: I thought that _may_ not work for directories but I didn't test it. I only tested `filetime` crate.

Comment (zanieb) [reply crates/install-wheel-rs/src/linker.rs]: https://github.com/astral-sh/uv/pull/2545/commits/205b224ac5a475618520c1ea8aeeff9703eb677b

File: crates/install-wheel-rs/src/linker.rs [hunk 4]
```diff
@@ -249,6 +250,33 @@ fn clone_wheel_files(
...
+
+        // `File.set_modified` is not available in `fs_err` yet
+        #[allow(clippy::disallowed_types)]
+        match std::fs::File::open(site_packages.as_ref()) {
+            Ok(dir) => {
+                if let Err(err) = dir.set_modified(now) {
+                    debug!(
+                        "Failed to update mtime for {}: {err}",
+                        site_packages.as_ref().display()
+                    );
```
Comment (zanieb) [inline crates/install-wheel-rs/src/linker.rs:270]: I'm a little unsure about this error handling but I'd prefer not to fail here?

File: crates/install-wheel-rs/src/linker.rs [hunk 5]
```diff
@@ -249,6 +250,30 @@ fn clone_wheel_files(
...
+    // import mechanisms to determine if it should look for new packages in a directory.
+    // Here, we force the mtime to be updated to ensure that packages are importable without
+    // manual cache invalidation.
+    //
+    // <https://github.com/python/cpython/blob/8336cb2b6f428246803b02a4e97fce49d0bb1e09/Lib/importlib/_bootstrap_external.py#L1601>
+    let now = SystemTime::now();
+
+    // `File.set_modified` is not available in `fs_err` yet
+    #[allow(clippy::disallowed_types)]
+    match std::fs::File::open(site_packages.as_ref()) {
```
Comment (zanieb) [inline crates/install-wheel-rs/src/linker.rs:263]: https://github.com/andrewhickman/fs-err/issues/54

File: crates/uv/tests/pip_install.rs [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
-use std::process::Command;
-
 use anyhow::Result;
 use assert_cmd::prelude::*;
 use assert_fs::prelude::*;
 use base64::{prelude::BASE64_STANDARD as base64, Engine};
 use indoc::indoc;
 use itertools::Itertools;
+#[cfg(unix)]
+use std::os::unix::fs::MetadataExt;
```
Comment (charliermarsh) [inline crates/uv/tests/pip_install.rs:10]: You could also put this import within the test -- we tend to do that in a few other places.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Nice, thanks!

**Comment by zanieb** [comment]
Interestingly the test failed on Ubuntu so I removed the macOS gate and we'll do this whenever we clone

**Comment by charliermarsh** [comment]
That's cool with me.

**Comment by zanieb** [comment]
Hm it seems to be flaky...

**Comment by charliermarsh** [comment]
The test, or the fix?

**Comment by zanieb** [comment]
Well both I guess :) neither the initial reproduction or the test have ever failed for me locally though.

We might want to bump `now` by `1` to make this particularly robust... I'm hesitant though. Hopefully moving it to after the clone itself will ensure the time isn't the same as when it started — thinking the flake might be resolution related?

**Comment by zanieb** [comment]
Hm seems like the test is still flaky.. will investigate further.

**Comment by zanieb** [comment]
It was just the test, `mtime()` is second-resolution but we're faster than that sometimes so we need to compare `mtime_nsec()`

**Pull request merged**